### PR TITLE
Implement calculations in periodic boundary conditions

### PR DESCRIPTION
**Pull request by gcassella**
Includes the feature layer, envelope, and Hamiltonian described in
```
Cassella, G., Sutterud, H., Azadi, S., Drummond, N.D., Pfau, D.,
Spencer, J.S. and Foulkes, W.M.C., 2022. Discovering Quantum Phase Transitions
with Fermionic Neural Networks. arXiv preprint arXiv:2202.05183.
```

Several changes to the configuration interface for feature layers and envelopes have been made. These are now specified by an interface similar to the local energy function. These changes, and the new periodic boundary conditions, are demonstrated in `ferminet.configs.heg`.

Other miscellaneous changes:
- Reduced failure threshold for the `pylint` step of the GitHub CI workflow to `9.75/10.0`. This prevents the CI from failing over linting errors that are acceptable under the in-house style.
- Implemented dummy element 'X' with atomic number zero. Used to define the coordinate system for one-electron features in systems without nuclei (e.g. the homogeneous electron gas).

### CODE REVIEW

**Review by jsspencer — APPROVED** [review]
This is awesome, thank you!

File: ferminet/pbc/hamiltonian.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    return (4 * jnp.pi / volume) * jnp.sum(
+        jnp.exp(1.0j * jnp.dot(rec_vectors, separation)) * \
+        jnp.exp(-rec_vec_square / (4 * gamma)) / rec_vec_square)
+
+  def ewald_sum(separation: jnp.ndarray):
+    """Evaluates combined real and reciprocal space Ewald potential."""
+    return real_sum(separation) + recp_sum(separation) - jnp.pi / (
+        volume * gamma)
+
+  lat_vec_norm = jnp.linalg.norm(lat_vectors[1:], axis=-1)
```
Comment (jsspencer) [inline ferminet/pbc/hamiltonian.py:89]: Would be more readable if you moved this above `real_sum` so all "top-level" code is grouped together, before the internal functions which close over the variables here.

File: ferminet/pbc/hamiltonian.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  # and this value is chosen to optimize the convergence trade-off between the
+  # two sums. See CASINO QMC manual.
+  gamma = (2.8 / volume**(1 / 3))**2
+  ordinals = sorted(range(-truncation_limit, truncation_limit + 1), key=abs)
+  ordinals = jnp.array(list(product(ordinals, repeat=3)))
+  lat_vectors = jnp.einsum('kj,ij->ik', lattice, ordinals)
+  rec_vectors = jnp.einsum('kj,ij->ik', rec, ordinals[1:])
+  rec_vec_square = jnp.einsum('ij,ij->i', rec_vectors, rec_vectors)
+
+  def real_sum(separation: jnp.ndarray):
```
Comment (jsspencer) [inline ferminet/pbc/hamiltonian.py:69]: A little more verbose name here

real_space_ewald? (similarly for recp_sum)

**Review by jsspencer — CHANGES_REQUESTED** [review]

File: ferminet/train.py [hunk 3]
```diff
@@ -406,14 +434,14 @@ def train(cfg: ml_collections.ConfigDict, writer_manager=None):
     local_energy_module, local_energy_fn = (
         cfg.system.make_local_energy_fn.rsplit('.', maxsplit=1))
     local_energy_module = importlib.import_module(local_energy_module)
-    make_local_energy = getattr(local_energy_module, local_energy_fn)  # type: hamiltonian.MakeLocalEnergy
```
[the remark this answers is not in the document]
Comment (gcassella) [reply ferminet/train.py:409]: This is actually correct - the previous version was wrong! Pytype agrees.

**Comment by gcassella** [comment]
I think that addresses all of the comments - thank you for taking the time to go through everything and make clear recommendations :smile:

**Review by jsspencer — APPROVED** [review]

**Pull request merged**